kazoo/tests/test_connection.py

```diff
@@ -100,32 +99,126 @@ def back(state):
 
     def test_connection_write_timeout(self):
         client = self.client
-        ev = threading.Event()
+        ev_suspended = threading.Event()
+        ev_can_connect = threading.Event()
+        ev_connected = threading.Event()
         path = "/" + uuid.uuid4().hex
         handler = client.handler
         _select = handler.select
-        _socket = client._connection._socket
+        _create_connection = handler.create_connection
+
+        def controlled_create_connection(*args, **kwargs):
+            ev_can_connect.wait(5)
```

Comment:
This feels like a really involved test setup. Would you not be able to reproduce the same with
 * ZkClient configured with infinite retries, flat backoff, zero jitter
 * ZkServer initially downed.

1. You start the client -> it enters the connecting state as it keeps trying to connect
2. You can submit commands -> no exceptions (with your patch)
3. You start the server -> the commands should be effected.

I think it would make the test case less intrusive.
What do you think?

Reply:
Argh :)  What you are suggesting is close to what the previous test was doing:

https://github.com/python-zk/kazoo/pull/583/files#diff-b526bcbf3d9f758b2032e6763cb453c7R601

(Except for the "starting with ZK down" bit, which would indeed make things easier.  I don't know if any other tests start/stop ZK, though; I will check.)

I have switched because `test_connection_write_timeout`, which was already using a pretty intrusive method, was close to what I wanted to test—and I thought that making a single test a bit more complicated was the lesser evil.

The new `test_connection_lost_empties_queue` test, which is largely a duplicate of `test_connection_write_timeout`, also verifies that the queue empties out when the session cannot be recovered.  Would you know of a trick to invalidate the session without being intrusive?  (Besides submitting `_SESSION_EXPIRED`, which is inapplicable because it goes through the queue.)

I must admit that the thing which bothers me more in this test file is the duplication, which I made even worse; I was thinking of factoring out at least some of these local functions, and to submit a subsequent PR if the result turned out to be cleaner.

Reply:
Hum, how about starting the client with a random ID, of course unknown to the server. Wouldn't that trigger a session expired error and clear the queues?

As for server up/down, I *think* I saw this in the tests pertaining to failovers. I might be wrong, would need a minute to go through the tests with a PC

Reply:
Okay—I have now moved by tests back to `TestClient`, and applied some of your suggestions:

Starting with ZK down is not easy, due to the way `ZookeeperCluster` the test suite interact, and the fact that `teardown_zookeeper` does not tear down the cluster :)  Moreover, the first state change a client perceives is `CONNECTED`; it never goes from the initial `LOST` to `SUSPENDED`.

So instead, I create a fresh client which is backed by a single server, and stop the latter to force a transition to `SUSPENDED`.

As for flushing: fudging the session password before reconnect is indeed a good way to trigger `SessionExpiredError`. (Surprisingly enough, just invoking `KazooClient._reset_session` isn't: a new session gets created, but the state directly transitions from `SUSPENDED` to `CONNECTED`!)

Moving the two tests back to `TestClient` means that they are also exercised under the `gevent` and `eventlet` handlers.